-
Notifications
You must be signed in to change notification settings - Fork 9
Expand file tree
/
Copy pathcuda_fim_kernel.cu
More file actions
431 lines (356 loc) · 9.99 KB
/
Copy pathcuda_fim_kernel.cu
File metadata and controls
431 lines (356 loc) · 9.99 KB
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
277
278
279
280
281
282
283
284
285
286
287
288
289
290
291
292
293
294
295
296
297
298
299
300
301
302
303
304
305
306
307
308
309
310
311
312
313
314
315
316
317
318
319
320
321
322
323
324
325
326
327
328
329
330
331
332
333
334
335
336
337
338
339
340
341
342
343
344
345
346
347
348
349
350
351
352
353
354
355
356
357
358
359
360
361
362
363
364
365
366
367
368
369
370
371
372
373
374
375
376
377
378
379
380
381
382
383
384
385
386
387
388
389
390
391
392
393
394
395
396
397
398
399
400
401
402
403
404
405
406
407
408
409
410
411
412
413
414
415
416
417
418
419
420
421
422
423
424
425
426
427
428
429
430
431
//
// CUDA implementation of FIM (Fast Iterative Method) for Eikonal equations
//
// Copyright (c) Won-Ki Jeong (wkjeong@unist.ac.kr)
//
// 2016. 2. 4
//
#include "cuda_fim_kernel.h"
__device__ DOUBLE get_time_eikonal(DOUBLE a, DOUBLE b, DOUBLE c, DOUBLE s)
{
DOUBLE ret, tmp;
// a > b > c
if(a < b) { tmp = a; a = b; b = tmp; }
if(b < c) { tmp = b; b = c; c = tmp; }
if(a < b) { tmp = a; a = b; b = tmp; }
ret = INF;
if(c < INF)
{
ret = c + s;
if(ret > b)
{
tmp = ((b+c) + sqrtf(2.0f*s*s-(b-c)*(b-c)))*0.5f;
if(tmp > b) ret = tmp;
if(ret > a) {
tmp = (a+b+c)/3.0f + sqrtf(2.0f*(a*(b-a)+b*(c-b)+c*(a-c))+3.0f*s*s)/3.0f;
if(tmp > a) ret = tmp;
}
}
}
return ret;
}
__global__ void run_solver(double* spd, bool* mask, const DOUBLE *sol_in, DOUBLE *sol_out, bool *con, uint* list, int xdim, int ydim, int zdim, int nIter, uint nActiveBlock)
{
uint list_idx = blockIdx.y*gridDim.x + blockIdx.x;
if(list_idx < nActiveBlock)
{
// retrieve actual block index from the active list
uint block_idx = list[list_idx];
double F;
bool isValid;
uint blocksize = BLOCK_LENGTH*BLOCK_LENGTH*BLOCK_LENGTH;
uint base_addr = block_idx*blocksize;
uint xgridlength = xdim/BLOCK_LENGTH;
uint ygridlength = ydim/BLOCK_LENGTH;
uint zgridlength = zdim/BLOCK_LENGTH;
// compute block index
uint bx = block_idx%xgridlength;
uint tmpIdx = (block_idx - bx)/xgridlength;
uint by = tmpIdx%ygridlength;
uint bz = (tmpIdx-by)/ygridlength;
uint tx = threadIdx.x;
uint ty = threadIdx.y;
uint tz = threadIdx.z;
uint tIdx = tz*BLOCK_LENGTH*BLOCK_LENGTH + ty*BLOCK_LENGTH + tx;
__shared__ DOUBLE _sol[BLOCK_LENGTH+2][BLOCK_LENGTH+2][BLOCK_LENGTH+2];
// copy global to shared memory
dim3 idx(tx+1,ty+1,tz+1);
SOL(idx.x,idx.y,idx.z) = sol_in[base_addr + tIdx];
//F = _speed[(int)spd[base_addr + tIdx]];
F = spd[base_addr + tIdx];
if(F > 0) F = 1.0/F; // F = 1/f
isValid = mask[base_addr + tIdx];
uint new_base_addr, new_tIdx;
// 1-neighborhood values
if(tx == 0)
{
if(bx == 0) // end of the grid
{
new_tIdx = tIdx;
new_base_addr = base_addr;
}
else
{
new_tIdx = tIdx + BLOCK_LENGTH-1;
new_base_addr = (block_idx - 1)*blocksize;
}
SOL(tx,idx.y,idx.z) = sol_in[new_base_addr + new_tIdx];
}
if(tx == BLOCK_LENGTH-1)
{
if(bx == xgridlength-1) // end of the grid
{
new_tIdx = tIdx;
new_base_addr = base_addr;
}
else
{
new_tIdx = tIdx - (BLOCK_LENGTH-1);
new_base_addr = (block_idx + 1)*blocksize;
}
SOL(tx+2,idx.y,idx.z) = sol_in[new_base_addr + new_tIdx];
}
if(ty == 0)
{
if(by == 0)
{
new_tIdx = tIdx;
new_base_addr = base_addr;
}
else
{
new_tIdx = tIdx + (BLOCK_LENGTH-1)*BLOCK_LENGTH;
new_base_addr = (block_idx - xgridlength)*blocksize;
}
SOL(idx.x,ty,idx.z) = sol_in[new_base_addr + new_tIdx];
}
if(ty == BLOCK_LENGTH-1)
{
if(by == ygridlength-1)
{
new_tIdx = tIdx;
new_base_addr = base_addr;
}
else
{
new_tIdx = tIdx - (BLOCK_LENGTH-1)*BLOCK_LENGTH;
new_base_addr = (block_idx + xgridlength)*blocksize;
}
SOL(idx.x,ty+2,idx.z) = sol_in[new_base_addr + new_tIdx];
}
if(tz == 0)
{
if(bz == 0)
{
new_tIdx = tIdx;
new_base_addr = base_addr;
}
else
{
new_tIdx = tIdx + (BLOCK_LENGTH-1)*BLOCK_LENGTH*BLOCK_LENGTH;
new_base_addr = (block_idx - xgridlength*ygridlength)*blocksize;
}
SOL(idx.x,idx.y,tz) = sol_in[new_base_addr + new_tIdx];
}
if(tz == BLOCK_LENGTH-1)
{
if(bz == zgridlength-1)
{
new_tIdx = tIdx;
new_base_addr = base_addr;
}
else
{
new_tIdx = tIdx - (BLOCK_LENGTH-1)*BLOCK_LENGTH*BLOCK_LENGTH;
new_base_addr = (block_idx + xgridlength*ygridlength)*blocksize;
}
SOL(idx.x,idx.y,tz+2) = sol_in[new_base_addr + new_tIdx];
}
__syncthreads();
DOUBLE a,b,c,oldT,newT;
for(int iter=0; iter<nIter; iter++)
{
//
// compute new value
//
oldT = newT = SOL(idx.x,idx.y,idx.z);
if(isValid)
{
a = min(SOL(tx,idx.y,idx.z),SOL(tx+2,idx.y,idx.z));
b = min(SOL(idx.x,ty,idx.z),SOL(idx.x,ty+2,idx.z));
c = min(SOL(idx.x,idx.y,tz),SOL(idx.x,idx.y,tz+2));
DOUBLE tmp = (DOUBLE) get_time_eikonal(a, b, c, F);
newT = min(tmp,oldT);
}
__syncthreads();
if(isValid) SOL(idx.x,idx.y,idx.z) = newT;
__syncthreads(); // this may not required
}
DOUBLE residue = oldT - newT;
// write back to global memory
con[base_addr + tIdx] = (residue < EPS) ? true : false;
sol_out[base_addr + tIdx] = newT;
}
}
__global__ void run_reduction(bool *con, bool *listVol, uint *list, uint nActiveBlock)
{
uint list_idx = blockIdx.y*gridDim.x + blockIdx.x;
if(list_idx < nActiveBlock)
{
uint block_idx = list[list_idx];
__shared__ bool conv[BLOCK_LENGTH*BLOCK_LENGTH*BLOCK_LENGTH];
uint blocksize = BLOCK_LENGTH*BLOCK_LENGTH*BLOCK_LENGTH/2;
uint base_addr = block_idx*blocksize*2;
uint tx = threadIdx.x;
uint ty = threadIdx.y;
uint tz = threadIdx.z;
uint tIdx = tz*BLOCK_LENGTH*BLOCK_LENGTH + ty*BLOCK_LENGTH + tx;
conv[tIdx] = con[base_addr + tIdx];
conv[tIdx + blocksize] = con[base_addr + tIdx + blocksize];
__syncthreads();
for(uint i=blocksize; i>0; i/=2)
{
if(tIdx < i)
{
bool b1, b2;
b1 = conv[tIdx];
b2 = conv[tIdx+i];
conv[tIdx] = (b1 && b2) ? true : false ;
}
__syncthreads();
}
if(tIdx == 0)
{
listVol[block_idx] = !conv[0]; // active list is negation of tile convergence (active = not converged)
}
}
}
__global__ void run_check_neighbor(double* spd, bool* mask, const DOUBLE *sol_in, DOUBLE *sol_out,
bool *con, uint* list, int xdim, int ydim, int zdim,
uint nActiveBlock, uint nTotalBlock)
{
uint list_idx = blockIdx.y*gridDim.x + blockIdx.x;
if(list_idx < nTotalBlock)
{
double F;
bool isValid;
__shared__ DOUBLE _sol[BLOCK_LENGTH+2][BLOCK_LENGTH+2][BLOCK_LENGTH+2];
uint block_idx = list[list_idx];
uint blocksize = BLOCK_LENGTH*BLOCK_LENGTH*BLOCK_LENGTH;
uint base_addr = block_idx*blocksize;
uint tx = threadIdx.x;
uint ty = threadIdx.y;
uint tz = threadIdx.z;
uint tIdx = tz*BLOCK_LENGTH*BLOCK_LENGTH + ty*BLOCK_LENGTH + tx;
if(list_idx < nActiveBlock) // copy value
{
sol_out[base_addr + tIdx] = sol_in[base_addr + tIdx];
}
else
{
uint xgridlength = xdim/BLOCK_LENGTH;
uint ygridlength = ydim/BLOCK_LENGTH;
uint zgridlength = zdim/BLOCK_LENGTH;
// compute block index
uint bx = block_idx%xgridlength;
uint tmpIdx = (block_idx - bx)/xgridlength;
uint by = tmpIdx%ygridlength;
uint bz = (tmpIdx-by)/ygridlength;
#ifdef __DEVICE_EMULATION__
assert(block_idx == bz*xgridlength*ygridlength + by*xgridlength + bx);
printf("Block %d's index : %d, %d, %d\n", block_idx, bx, by, bz);
printf("Thread %d's index : %d, %d, %d\n", tIdx, tx, ty, tz);
#endif
// copy global to shared memory
dim3 idx(tx+1,ty+1,tz+1);
_sol[idx.x][idx.y][idx.z] = sol_in[base_addr + tIdx];
F = spd[base_addr + tIdx];
if(F > 0) F = 1.0/F;
isValid = mask[base_addr + tIdx];
uint new_base_addr, new_tIdx;
// 1-neighborhood values
if(tx == 0)
{
if(bx == 0) // end of the grid
{
new_tIdx = tIdx;
new_base_addr = base_addr;
}
else
{
new_tIdx = tIdx + BLOCK_LENGTH-1;
new_base_addr = (block_idx - 1)*blocksize;
}
_sol[tx][idx.y][idx.z] = sol_in[new_base_addr + new_tIdx];
}
if(tx == BLOCK_LENGTH-1)
{
if(bx == xgridlength-1) // end of the grid
{
new_tIdx = tIdx;
new_base_addr = base_addr;
}
else
{
new_tIdx = tIdx - (BLOCK_LENGTH-1);
new_base_addr = (block_idx + 1)*blocksize;
}
_sol[tx+2][idx.y][idx.z] = sol_in[new_base_addr + new_tIdx];
}
if(ty == 0)
{
if(by == 0)
{
new_tIdx = tIdx;
new_base_addr = base_addr;
}
else
{
new_tIdx = tIdx + (BLOCK_LENGTH-1)*BLOCK_LENGTH;
new_base_addr = (block_idx - xgridlength)*blocksize;
}
_sol[idx.x][ty][idx.z] = sol_in[new_base_addr + new_tIdx];
}
if(ty == BLOCK_LENGTH-1)
{
if(by == ygridlength-1)
{
new_tIdx = tIdx;
new_base_addr = base_addr;
}
else
{
new_tIdx = tIdx - (BLOCK_LENGTH-1)*BLOCK_LENGTH;
new_base_addr = (block_idx + xgridlength)*blocksize;
}
_sol[idx.x][ty+2][idx.z] = sol_in[new_base_addr + new_tIdx];
}
if(tz == 0)
{
if(bz == 0)
{
new_tIdx = tIdx;
new_base_addr = base_addr;
}
else
{
new_tIdx = tIdx + (BLOCK_LENGTH-1)*BLOCK_LENGTH*BLOCK_LENGTH;
new_base_addr = (block_idx - xgridlength*ygridlength)*blocksize;
}
_sol[idx.x][idx.y][tz] = sol_in[new_base_addr + new_tIdx];
}
if(tz == BLOCK_LENGTH-1)
{
if(bz == zgridlength-1)
{
new_tIdx = tIdx;
new_base_addr = base_addr;
}
else
{
new_tIdx = tIdx - (BLOCK_LENGTH-1)*BLOCK_LENGTH*BLOCK_LENGTH;
new_base_addr = (block_idx + xgridlength*ygridlength)*blocksize;
}
_sol[idx.x][idx.y][tz+2] = sol_in[new_base_addr + new_tIdx];
}
__syncthreads();
DOUBLE a,b,c,oldT,newT;
//
// compute new value
//
oldT = newT = _sol[idx.x][idx.y][idx.z];
if(isValid)
{
a = min(_sol[tx][idx.y][idx.z],_sol[tx+2][idx.y][idx.z]);
b = min(_sol[idx.x][ty][idx.z],_sol[idx.x][ty+2][idx.z]);
c = min(_sol[idx.x][idx.y][tz],_sol[idx.x][idx.y][tz+2]);
DOUBLE tmp = (DOUBLE) get_time_eikonal(a, b, c, F);
newT = min(tmp,oldT);
sol_out[base_addr + tIdx] = newT;
}
// write back to global memory
DOUBLE residue = oldT - newT;
con[base_addr + tIdx] = (residue < EPS) ? true : false;
}
}
}